autoprecompiles/src/equivalence_classes.rs

```diff
@@ -0,0 +1,29 @@
+use std::collections::BTreeSet;
+
+use serde::{Deserialize, Serialize};
+
+/// An equivalence class, i.e, a set of values of type `T` which are considered equivalent
+pub type EquivalenceClass<T> = BTreeSet<T>;
+
+/// A collection of equivalence classes where all classes are guaranteed to have at least two elements
+/// This is enforced by construction of this type only happening through collection, where we ignore empty and singleton classes
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+#[serde(bound(deserialize = "T: Ord + Deserialize<'de>"))]
+pub struct EquivalenceClasses<T> {
```

Comment:
I had named this `Partition` in 42cf7c725bec098bfbe57b59f1e5a223907561e7, sticking to that name :)